### PR TITLE
Feature complete user profile

### PR DESCRIPTION
**Pull request by The-Vector-007**

### CODE REVIEW

**Comment by arpansac** [comment]
Task linked: [CU-f1u25t User sign up journey revamp](https://app.clickup.com/t/f1u25t)

File: projects/commudle-admin/src/app/app.component.html [hunk 1]
```diff
@@ -35,6 +35,7 @@
   </nb-sidebar>
 
   <nb-layout-column>
+    <app-profile-status-bar *ngIf="currentUser !== null && profileBarStatus"></app-profile-status-bar>
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/app.component.html]: Do we need to check null validity explicitly?

File: projects/commudle-admin/src/app/services/stepper.service.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      if (this.currentUser) {
+        this.calculateProfilePercentage();
+      }
+    });
+  }
+
+  calculateProfilePercentage() {
+    let profilePercentage = 0;
+
+    if (this.currentUser.tags.length >= 5) {
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/services/stepper.service.ts:57]: The tags will never be greater than 5, so this condition can be improved.

Also can we give weightage to each tag they have filled (like 4% for each)

File: projects/commudle-admin/src/app/services/stepper.service.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    for (const field in this.profileWeights) {
+      if (this.currentUser[field]) {
+        profilePercentage += this.profileWeights[field];
+      }
+    }
+    this.profileCompletePercentage.next(profilePercentage);
+  }
+
+  showStepper() {
+    this.dialogRef = this.dialogService.open(StepperComponent, { hasScroll: true, closeOnBackdropClick: false });
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/services/stepper.service.ts]: Let's disable close on pressing `esc` key also

File: projects/commudle-admin/src/app/services/stepper.service.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { LibAuthwatchService } from 'projects/shared-services/lib-authwatch.service';
+import { BehaviorSubject } from 'rxjs';
+import { StepperComponent } from 'projects/commudle-admin/src/app/components/stepper/stepper.component';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class StepperService {
+  // user basic info = 70% , total fields = 7, each weight = 10
+  // social links = 30%, total fields = 10, each weight = 3
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/services/stepper.service.ts]: change this comment or remove it

File: projects/commudle-admin/src/app/services/stepper.service.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    name: 10,
+    designation: 10,
+    about_me: 10,
+    location: 10,
+    gender: 10,
+    skills: 20,
+    socialLinks: 10,
+  };
+
+  private socialLinks = [
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/services/stepper.service.ts:24]: do we need a separate array for the social links? can we merge it into the object above?

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        this.usernameForm.patchValue({ username: this.currentUser.username });
+      }
+    });
+
+    this.userProfileManagerService.updateUsername$.subscribe((value) => {
+      if (value) {
+        if (!this.router.url.includes('/users/' + this.lastUsername)) {
+          this.reloadPage = false;
+        }
+        if (this.currentUsername != this.lastUsername) {
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.ts]: Let's use `!==`

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.html [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<div class="username">
+  <form (ngSubmit)="confirmSubmissionDialogueOpen()" [formGroup]="usernameForm">
+    <div>
+      <label>
+        Choose a username to for your public profile
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.html]: Let's change the text here

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/social-links/social-links.component.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+@Component({
+  selector: 'app-social-links',
+  templateUrl: './social-links.component.html',
+  styleUrls: ['./social-links.component.scss'],
+})
+export class SocialLinksComponent implements OnInit {
+  currentUser: ICurrentUser;
+  @Output() socialLinksFormValidity: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  socialLinksForm = this.fb.group({
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/social-links/social-links.component.ts:17]: Let's add some regex validations for the links

File: projects/commudle-admin/src/app/components/stepper/stepper.component.scss [hunk 9]
```diff
@@ -0,0 +1,114 @@
+@import './projects/commudle-admin/src/assets/styles/variables';
+
+nb-card {
+  overflow: hidden;
+  height: 97vh;
+  width: 800px;
+
+  @media all and (max-width: $tab-width) {
```
Comment (Ashwin-op) [inline projects/commudle-admin/src/app/components/stepper/stepper.component.scss:8]: Let's not do this

**Review by Ashwin-op — CHANGES_REQUESTED** [review]

**Comment by Ashwin-op** [comment]
Bug fixes

- Next button should be disabled when the username is blank
image

- Next button should be disabled when the skills are blank
image

- The navigation buttons should be at the bottom
image

- Let's also show `username available` here
image

File: projects/commudle-admin/src/app/app.component.html (hunk 1 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/app.component.html]: removed

File: projects/commudle-admin/src/app/services/stepper.service.ts (hunk 3 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/services/stepper.service.ts]: done

File: projects/commudle-admin/src/app/services/stepper.service.ts (hunk 4 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/services/stepper.service.ts]: removed

File: projects/commudle-admin/src/app/services/stepper.service.ts (hunk 5 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/services/stepper.service.ts:24]: we need to keep it separate.

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.ts (hunk 6 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.ts]: done

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.html (hunk 7 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/username/username.component.html]: done

File: projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/social-links/social-links.component.ts (hunk 8 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/feature-modules/users/components/public-profile/user-basic-details/basic-user-profile/social-links/social-links.component.ts:17]: we are using a library which takes the username and convert it into a valid url automatically

File: projects/commudle-admin/src/app/components/stepper/stepper.component.scss (hunk 9 above)
Comment (The-Vector-007) [reply projects/commudle-admin/src/app/components/stepper/stepper.component.scss:8]: done

**Pull request merged**